File: src/my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

class NumberCounter : public rclcpp::Node
{
public:
    NumberCounter() : Node("number_counter")
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10,
            std::bind(&NumberCounter::callbackNumberPublisher, this, std::placeholders::_1));

        counter_ = 0;
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);

        server_ = this->create_service<example_interfaces::srv::SetBool>("reset_counter", bind(&NumberCounter::callbackResetCounter, this, std::placeholders::_1, std::placeholders::_2));
        
        RCLCPP_INFO(this->get_logger(), "Number Counter has been started");
    }

private:
    void callbackNumberPublisher(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        counter_ += msg->data;
        RCLCPP_INFO(this->get_logger(), "Received from number: %d", msg->data);

        auto msg_count = example_interfaces::msg::Int64();
        msg_count.data = this->counter_;
        publisher_->publish(msg_count);
    }

    void callbackResetCounter(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                              const example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if(request->data){
            counter_ = 0;
            response->success = true;
            response->message = "counter reset to 0";
        }
    }

    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
    int counter_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    //shared pointer, you don't need to delete manually the instance when object is not referenced (like GC in Java)
    auto node = std::make_shared<NumberCounter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool

class NumberCounter(Node):

    def __init__(self):
        super().__init__("number_counter") 

        self.counter_ = 0;

        self.subscriber_ = self.create_subscription(Int64, "number", self.callback_number_publisher, 10)

        self.publisher_ = self.create_publisher(Int64, "number_count", 10) # create topic number_count in which number_counter will publish the count
        
        self.server_ = self.create_service(SetBool, "reset_counter", self.callback_reset_counter)

        self.get_logger().info("Number Counter has been started")
    

    def callback_number_publisher(self, msg):
        self.counter_ += msg.data
        self.get_logger().info("Received from /number: {}".format(msg.data))
        msg = Int64()
        msg.data = self.counter_
        self.publisher_.publish(msg)

    def callback_reset_counter(self, request, response):
        if(request.data):
            self.counter_ = 0
            response.success = True
            response.message = "Counter reset to 0"
        return response

def main(args=None):
    rclpy.init(args=args) # first line to be written in any ROS2 .py node
    node = NumberCounter()
    rclpy.spin(node) # allow node to continue to be alive
    rclpy.shutdown() # last line of any ROS2 .py node

if __name__ == "__main__":
    main()